### ISSUE TITLE
[TODO] [Day 4] WanDB 연동 & baseline 업데이트

### ISSUE THREAD

**Issue by nimnusmik** [AIBootcamp14/computervisioncompetition-cv-1#20]
## 📋 작업 개요
**작업 주제:** WanDB 연동
**담당자:** @nimnusmik   
**마감일:** 9/3 12:00  
**예상 소요시간:** 총 3시간 

## 📅 기간
- 시작일: 2025-09-04
- 종료일: 2025-09-04

## 📌 목표
- 베이스라인 코드에 WandB 연동하여 실험 추적 시스템 구축
- 코드 배포 및 사용법 공유

### 🔧 WandB 기본 설정
- [x] **WandB 계정 생성 및 API 키 설정**
  - WandB 팀 프로젝트 생성: `document-classification-team`
  - 팀원 초대 및 권한 설정
  - API 키 환경변수 설정 (`.env` 파일)

- [x] **기본 라이브러리 설치 및 import 추가**
 
### 📊 실험 추적 로직 구현

- [x] **프로젝트 초기화 코드 작성**
  - 프로젝트명, 실험명 자동 생성 로직
  - Config 파라미터 정의 (LR, BATCH_SIZE, model_name, img_size 등)
  - Tags 시스템 설계 (model_type, fold_num, experiment_type)

- [ ] **하이퍼파라미터 로깅**
  - 모든 설정값 wandb.config에 저장
  - Optuna 최적화 결과 연동
  - 데이터 분할 정보 (train/val split ratio)

- [ ] **K-Fold CV 결과 추적**
  - 각 Fold별 개별 run 생성
  - Fold별 train/val loss, accuracy, f1_score 실시간 로깅
  - Cross-validation summary 테이블 생성

- [ ] **학습 과정 실시간 모니터링**
  - Epoch별 metrics 로깅 (train_loss, train_acc, train_f1, val_loss, val_acc, val_f1)
  - Learning rate 스케줄링 추적
  - GPU 사용률, 메모리 사용량 모니터링
  - 학습 시간 측정 및 로깅

### 📈 시각화 및 분석

- [ ] **Confusion Matrix 시각화**
  - 각 Fold별 confusion matrix 생성
  - 클래스별 성능 분석 차트
  - 최종 앙상블 결과 confusion matrix

- [ ] **모델 성능 비교 차트**
  - Fold별 성능 비교 bar chart
  - Loss curve 시각화 (train vs validation)
  - F1-score progression 차트

- [ ] **데이터 분석 시각화**
  - 클래스 분포 차트
  - 이미지 샘플 로깅 (각 클래스별 대표 이미지)
  - 증강된 이미지 샘플 비교

### 🔄 TTA 및 앙상블 추적

- [ ] **TTA 실험 로깅**
  - TTA transform별 성능 기여도 분석
  - 신뢰도 분포 히스토그램
  - TTA 전후 성능 비교 테이블

- [ ] **앙상블 결과 분석**
  - 5-fold 모델별 예측 일치도 분석
  - 앙상블 가중치 최적화 결과
  - 최종 submission 성능 예측

### 🎯 모델 아티팩트 관리

- [ ] **모델 체크포인트 저장**
  - 각 Fold 최고 성능 모델 wandb.save()
  - 모델 버전 관리 (model registry 활용)
  - 재현 가능한 체크포인트 관리

- [ ] **실험 결과물 저장**
  - 최종 prediction 파일 (choice.csv)
  - K-fold 교차검증 상세 결과
  - 실험 설정 파일 (.yaml 형태)

### 👥 팀 협업 환경 구축

- [ ] **코드 템플릿 제작**
  - 팀원용 simplified baseline 코드 작성
  - WandB 설정 가이드 문서 작성
  - 실험 네이밍 컨벤션 정의

- [ ] **팀 대시보드 구성**
  - 팀 전체 실험 비교 대시보드
  - 리더보드 스타일 성능 순위표
  - 실험 진행 상황 추적 차트

## 💡 참고 사항
- 현재 베이스라인: EfficientNet-B3 (384x384), 5-fold CV, TTA 적용
- 평가 지표: Macro F1-score (대회 공식 지표)
- 데이터: 17개 문서 타입 분류 문제

### 유용한 링크
- [ ] WandB 공식 문서: [https://docs.wandb.ai/](https://docs.wandb.ai/)
- [ ] PyTorch + WandB 예제: [https://wandb.ai/site/articles/intro-to-pytorch-with-wandb](https://wandb.ai/site/articles/intro-to-pytorch-with-wandb)
- [ ] K-Fold CV WandB 예제: [https://wandb.ai/wandb_fc/tips/reports/How-to-Use-Cross-Validation-in-W-B](https://wandb.ai/wandb_fc/tips/reports/How-to-Use-Cross-Validation-in-W-B)
- [ ] Timm 모델 로깅: [https://wandb.ai/geekyrakshit/timm-finetune](https://wandb.ai/geekyrakshit/timm-finetune)

---

### 실험 중 발견사항
-  

### 다음 실험 아이디어
- [ ] augmentation 기법 추가 (Mixup 30%, Cutout 30%, None 40%)
- [ ] 데이터 트랜스포메이션 추가 
 

### 팀원 피드백
-

## ⚠️ 주의사항
- API 키 노출 방지 (`.gitignore`에 `.env` 추가)
- 대용량 이미지 로깅 시 wandb 용량 제한 확인
- K-Fold 실험 시 너무 많은 run 생성 방지 → sweep 기능 활용 고려

**Comment by nimnusmik** [comment]
### 로그 분석: F1 최고 성능을 달성했지만…

* **성공적인 학습:** Epoch 25에서 **Val F1 0.9265**라는 높은 최고 성능을 달성했습니다. 이는 모델이 Fold 5의 검증 데이터셋에서 매우 뛰어난 일반화 능력을 보였다는 뜻입니다.
* **조기 종료 실패:** `patience`가 5로 설정되었음에도 불구하고, 학습이 **총 30 Epoch를 모두 채우고 종료**되었습니다. 이는 Epoch 25 이후 Val F1이 최고 성능보다 낮게 나오긴 했지만, 5번 연속으로 하락하지는 않았다는 의미입니다.
* **성능 정체:** 로그를 자세히 보면 Epoch 25 이후의 Val F1은 **`0.9144 -> 0.9199 -> 0.9191 -> 0.9234 -> 0.9224`**로, 하락과 상승을 반복하며 **성능이 정체되거나 미세하게 하락하는 모습**을 보입니다.

이러한 상황은 모델이 최고 성능에 도달한 후, 더 이상 유의미한 개선을 보이지 않고 있다는 신호입니다. 즉, 조기 종료 시점이 늦게 왔거나, 학습을 더 오래 진행했으면 더 좋은 성능을 찾을 수도 있었다는 가능성을 시사합니다.

---

### 다음 단계: 개선을 위한 전략

이러한 분석을 토대로, 모델의 성능을 더욱 확실하게 끌어올리기 위한 다음 단계는 다음과 같습니다.

### 1. Epoch 및 Patience 값 조정 📈

모델이 Epoch 30에서 멈췄다는 것은 학습을 더 오래 진행했어도 괜찮았을 수 있다는 의미입니다.

* **Epoch 수 늘리기:** 현재 30 Epoch를 50 또는 100으로 늘려보세요. 이는 모델이 성능의 최고점을 충분히 탐색할 시간을 줍니다.
* **Patience 값 늘리기:** `patience`를 5에서 **7~10** 정도로 늘려보세요. 이렇게 하면 모델이 최고 성능을 찍은 후, 일시적인 성능 하락에도 불구하고 더 안정적으로 최적의 종료 시점을 찾게 됩니다.

### 2. 모든 K-폴드 결과 분석 📊

지금까지는 **Fold 5**의 결과만 보셨습니다. 이 한 폴드의 결과는 모델의 전체적인 성능을 대변하지 못합니다.

* **전체 평균 확인:** 모든 5개의 폴드에 대한 최고 **Val F1 점수들의 평균**과 **표준편차**를 확인하세요.
* **결과 해석:**
    * **평균이 높을수록:** 모델의 일반적인 성능이 우수하다는 뜻입니다.
    * **표준편차가 낮을수록:** 각 폴드 간의 성능 차이가 적어 모델이 특정 데이터 분할에 덜 민감하고 **안정적**이라는 뜻입니다.

### 3. 데이터 증강 전략 재검토 🖼️

이번 로그에서도 Epoch 24에서 **Mixup**이 적용된 후 성능이 소폭 상승하기도 했지만, 최고점 이후에는 성능이 정체되었습니다.

* **후반부 증강 강도 낮추기:** 20 Epoch 이후에는 Mixup과 같이 강한 정규화 기법을 제외하고, **`Cutout`이나 `RandomCrop`만 사용하여** 모델이 미세한 부분까지 학습하도록 유도해 보세요.

---

**Comment by nimnusmik** [comment]
학습 할 가능성이 더 남아있어 Epoch와 patient 늘리는 방향으로 
성능 업데이트 (0.917 -> 0.927) &  괴리 줄이기 성공


<img width="604" height="190" alt="Image" src="https://github.com/user-attachments/assets/7022d21e-7d6d-4871-ae28-0d1cd454ebfe" />
<img width="1023" height="131" alt="Image" src="https://github.com/user-attachments/assets/352ba73c-d9fb-457b-ad9f-fa5675f3ea8f" />

**Comment by nimnusmik** [comment]
0 | account_number | 계좌번호
-- | -- | --
1 | application_for_payment_of_pregnant | 임신 관련 지급 신청서
2 | car_dashboard | 자동차 대시보드
3 | confirmation_of_admission_and_disc | 입학 및 등록 확인서
4 | diagnosis | 진단서
5 | driver_lisence | 운전면허증
6 | medical_bill _receipts | 진료비
7 | medical_outpatient_certificate | 영수증외래 진료 확인서
8 | national_id_card | 주민등록증 / 신분증
9 | passport | 여권
10 | payment_confirmation | 결제 확인서
11 | pharmaceutical_receipt | 약국 영수증
12 | prescription | 처방전
13 | resume | 이력서
14 | statement_of_opinion | 의견서 / 진술서
15 | vehicle_registration _certificate | 자동차 등록증
16 | vehicle_registration_plate | 자동차 번호판

### wandb 추적 결과 3,7,14 클래스를 잘 못맞추는 중 


<img width="1148" height="613" alt="Image" src="https://github.com/user-attachments/assets/b3641407-4c66-41d6-9ad0-10a6a7b45d35" />

3,7,14에 대한 데이터 보정을 해줘볼까??

**Comment by Bkankim** [comment]
너무 대단하고 훌륭한 인사이트네요 선민님!
데이터 보정 결과도 기대하겠습니다
